### ISSUE TITLE
[tune] Combine bayesianOptSearch with ray.tune random search

### ISSUE THREAD

**Issue by JiayuanSternLi** [ray-project/ray#6195]
### System information
- **Linux Ubuntu 16.04**:
- **Ray installed from  binary**:
- **Ray version 0.8.0dev6**:
- **Python version 3.7.4**:
- **Exact command to reproduce**:

### Describe the problem
I failed to combine bayesianOptSearch for continues parameters and the random search from ray.tune for discrete parameters, but one works while the other one is commented in my code. The code is based on the class-based example.

### Source code / logs
`
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

import argparse
import numpy as np
from tensorflow.keras.layers import Dense, Flatten, Conv2D
from tensorflow.keras import Model
from tensorflow.keras.datasets.mnist import load_data
from ray.tune.schedulers import AsyncHyperBandScheduler
from ray.tune.suggest.bayesopt import BayesOptSearch

from ray import tune

MAX_TRAIN_BATCH = 10
FROTN_END_CALL = False

parser = argparse.ArgumentParser()
parser.add_argument(
    "--smoke-test", action="store_true", help="Finish quickly for testing")
args, _ = parser.parse_known_args()

class MyModel(Model):
    def __init__(self, hiddens=128):
        super(MyModel, self).__init__()
        self.conv1 = Conv2D(32, 3, activation="relu")
        self.flatten = Flatten()
        self.d1 = Dense(hiddens, activation="relu")
        self.d2 = Dense(10, activation="softmax")

    def call(self, x):
        x = self.conv1(x)
        x = self.flatten(x)
        x = self.d1(x)
        return self.d2(x)

class MNISTTrainable(tune.Trainable):
    def _setup(self, config):
        # IMPORTANT: See the above note.
        import tensorflow as tf
        (x_train, y_train), (x_test, y_test) = load_data()
        x_train, x_test = x_train / 255.0, x_test / 255.0

        # Add a channels dimension
        x_train = x_train[..., tf.newaxis]
        x_test = x_test[..., tf.newaxis]
        self.train_ds = tf.data.Dataset.from_tensor_slices((x_train, y_train))
        self.train_ds = self.train_ds.shuffle(10000).batch(
            config.get("batch", 32))
        self.test_ds = tf.data.Dataset.from_tensor_slices((x_test,
                                                           y_test)).batch(32)

        self.model = MyModel(hiddens=config.get("hiddens", 128))
        self.loss_object = tf.keras.losses.SparseCategoricalCrossentropy()
        self.optimizer = tf.keras.optimizers.SGD()

        self.optimizer.learning_rate = config.get("lr", 0.01)
        self.optimizer.momentum = config.get("momentum", 0)
        self.optimizer.nesterov = config.get("nesterov", False)
        self.train_loss = tf.keras.metrics.Mean(name="train_loss")
        self.train_accuracy = tf.keras.metrics.SparseCategoricalAccuracy(
            name="train_accuracy")

        self.test_loss = tf.keras.metrics.Mean(name="test_loss")
        self.test_accuracy = tf.keras.metrics.SparseCategoricalAccuracy(
            name="test_accuracy")

        @tf.function
        def train_step(images, labels):
            with tf.GradientTape() as tape:
                predictions = self.model(images)
                loss = self.loss_object(labels, predictions)
            gradients = tape.gradient(loss, self.model.trainable_variables)
            self.optimizer.apply_gradients(
                zip(gradients, self.model.trainable_variables))

            self.train_loss(loss)
            self.train_accuracy(labels, predictions)

        @tf.function
        def test_step(images, labels):
            predictions = self.model(images)
            t_loss = self.loss_object(labels, predictions)

            self.test_loss(t_loss)
            self.test_accuracy(labels, predictions)

        self.tf_train_step = train_step
        self.tf_test_step = test_step

    def _train(self):
        self.train_loss.reset_states()
        self.train_accuracy.reset_states()
        self.test_loss.reset_states()
        self.test_accuracy.reset_states()

        for idx, (images, labels) in enumerate(self.train_ds):
            if idx > MAX_TRAIN_BATCH:  # This is optional and can be removed.
                break
            self.tf_train_step(images, labels)

        for test_images, test_labels in self.test_ds:
            self.tf_test_step(test_images, test_labels)

        # It is important to return tf.Tensors as numpy objects.
        return {
            "epoch": self.iteration,
            "loss": self.train_loss.result().numpy(),
            "accuracy": self.train_accuracy.result().numpy() * 100,
            "test_loss": self.test_loss.result().numpy(),
            "mean_accuracy": self.test_accuracy.result().numpy() * 100
        }

if __name__ == "__main__":
    load_data()  # we download data on the driver to avoid race conditions.
    
    space = {'lr': (0.001, 0.5)}
    algo = BayesOptSearch(
        space,
        max_concurrent=1,
        metric="mean_accuracy",
        mode="max",
        utility_kwargs={
            "kind": "ucb",
            "kappa": 2.5,
            "xi": 0.0
        })

    scheduler = AsyncHyperBandScheduler(metric="acc", mode="max")
    analysis = tune.run(
        MNISTTrainable,
        stop={"training_iteration": 5 if args.smoke_test else 100},
        search_alg = algo,  # comment 1
        verbose=1,
        num_samples = 1,
        scheduler=scheduler,
        config={
                "hiddens": tune.choice([32, 64, 128]) # comment 2
               },
        resources_per_trial={'gpu': 8, 'cpu': 40},
        name = "example-experiment"
        )

    print("Best config is", analysis.get_best_config(metric="mean_accuracy"))
    trials = analysis.trials
    print("All trials inform:", trials)
`

Hey, take a look at the line with comment "comment 1 and comment 2". Commenting one makes the other one work, but it doesn't work if both exist. Errors is attached belowed:
![image](https://user-images.githubusercontent.com/33815430/69136450-d7ae2180-0af5-11ea-9715-708e9387157a.png)

**Comment by hhbyyh** [comment]
Maybe check: https://github.com/ray-project/ray/issues/6167

**Comment by richardliaw** [comment]
Generally, you cannot mix `tune` primitives with a search algorithm. You'll have to specify what you're trying to do in the search algorithm search space.

**Comment by JiayuanSternLi** [comment]
Thanks, Richardliaw, I have checked the post of BayesianOptSearch but unfortunately, he says: "There is no principled way of dealing with discrete parameters using this package." In this case, do you have any idea that the combination of the random search for discrete parameters and bayesian based search for continues parameters??

**Comment by richardliaw** [comment]
awesome!
